server/channels/app/imports/import_validators.go

```diff
@@ -650,3 +654,37 @@ func isValidEmailBatchingInterval(emailInterval string) bool {
 		emailInterval == model.PreferenceEmailIntervalFifteen ||
 		emailInterval == model.PreferenceEmailIntervalHour
 }
+
+// isValidGuestRoles checks if the user has both guest roles in the same team or channel.
+// at this point we assume that the user has a valid role scheme.
+func isValidGuestRoles(data UserImportData) bool {
+	var isSystemGuest bool
+	if data.Roles != nil {
+		isSystemGuest = model.IsInRole(*data.Roles, model.SystemGuestRoleId)
+	}
+
+	var isTeamGuest, isChannelGuest bool
+	if data.Teams != nil {
+		for _, team := range *data.Teams {
+			if team.Roles != nil {
+				isTeamGuest = model.IsInRole(*team.Roles, model.TeamGuestRoleId)
```

Comment:
What if the user is part of two teams, and a guest in one team and not in the other? Then the variable will become true, and then false again in the next iteration. Is that intended? Same for channels.

Reply:
From the ticket, what I understood is a guest is always a guest.

> In Mattermost, guests are always guests. That is, if they are a guest in one channel or team, they are a guest in all channels and teams. Unfortunately, we don’t have a single source of truth when it comes to guests. That is…they must be guests at the System Level, guest at the Team Level and guests at the Channel level.

cc @sbishel

Reply:
Yes the role for any given entity (system level, team or channel level) has to be guest if a user is a guest. So the fix for the logic can be also made to check if there is inconsistency across teams/channels, will take a look into that. Good catch!